### ISSUE TITLE
SPDK nvmf target hung up or crashed while the updated test/nvmf/filesystem/filesystem.sh

### ISSUE THREAD

**Issue by liang2yan** [spdk/spdk#251]
The script content is as below:
`#!/usr/bin/env bash

testdir=$(readlink -f $(dirname $0))
rootdir=$(readlink -f $testdir/../../..)
source $rootdir/test/common/autotest_common.sh
source $rootdir/test/nvmf/common.sh

KERNEL_DIR="/home/kernel.tar.gz"
KERNEL_BASE=$(basename $KERNEL_DIR)
rpc_py="python $rootdir/scripts/rpc.py"

function trap_handle() {
#       nvme disconnect -n "nqn.2016-06.io.spdk:cnode1" || true
        ps -p $nvmfpid
        retval=$?
        if [ $retval -eq 0 ]; then
                $rpc_py delete_bdev lvs0/lbd_0
                $rpc_py destroy_lvol_store -l lvs0
        fi
}

set -e

RDMA_IP_LIST=$(get_available_rdma_ips)
NVMF_FIRST_TARGET_IP=$(echo "$RDMA_IP_LIST" | head -n 1)
if [ -z $NVMF_FIRST_TARGET_IP ]; then
        echo "no NIC for nvmf test"
        exit 0
fi

timing_enter fs_test
timing_enter start_nvmf_tgt

$rootdir/scripts/gen_nvme.sh >> $testdir/../nvmf.conf
# Start up the NVMf target in another process
$NVMF_APP -c $testdir/../nvmf.conf &
nvmfpid=$!

trap "killprocess $nvmfpid; exit 1" SIGINT SIGTERM EXIT

waitforlisten $nvmfpid
timing_exit start_nvmf_tgt

modprobe -v nvme-rdma

ls_guid="$($rpc_py construct_lvol_store Nvme0n1 lvs0 -c 1048576)"
free_mb=$(get_lvs_free_mb "$ls_guid")
lb_bdevs="$($rpc_py construct_lvol_bdev -u $ls_guid lbd_0 "$free_mb")"
$rpc_py construct_nvmf_subsystem nqn.2016-06.io.spdk:cnode1 "trtype:RDMA traddr:$NVMF_FIRST_TARGET_IP trsvcid:$NVMF_PORT" '' -a -s SPDK001 -n "$lb_bdevs"
nvme connect -t rdma -n "nqn.2016-06.io.spdk:cnode1" -a "$NVMF_FIRST_TARGET_IP" -s "$NVMF_PORT"

trap "trap_handle; killprocess $nvmfpid; exit 1" SIGINT SIGTERM EXIT

sleep 2
mkdir -p /mnt/device

devs=`lsblk -l -o NAME | grep nvme`

for dev in $devs; do
        timing_enter parted
        parted -s /dev/$dev mklabel msdos  mkpart primary '0%' '100%'
        timing_exit parted
        sleep 1

        for fstype in "ext4" "btrfs" "xfs"; do
                timing_enter $fstype
                if [ $fstype = ext4 ]; then
                        force=-F
                else
                        force=-f
                fi

                mkfs.${fstype} $force /dev/${dev}p1

                mount /dev/${dev}p1 /mnt/device

                if [ $RUN_NIGHTLY -eq 1 ]; then
                        cp -r ${KERNEL_DIR} /mnt/device
                        make -C /mnt/device/${KERNEL_BASE} clean
                        make -C /mnt/device/${KERNEL_BASE} -j64
                        rm -fr /mnt/device/${KERNEL_BASE}
                        umount /mountnt/device
                else
                        touch /mnt/device/aaa
                        sync
                        rm /mnt/device/aaa
                        sync
                        umount /mnt/device
                        timing_exit $fstype
                fi
        done

        parted -s /dev/$dev rm 1
done

sync

nvme disconnect -n "nqn.2016-06.io.spdk:cnode1" || true
$rpc_py delete_nvmf_subsystem nqn.2016-06.io.spdk:cnode1
$rpc_py delete_bdev  "lvs0/lbd_0"
$rpc_py destroy_lvol_store -l lvs0

trap - SIGINT SIGTERM EXIT

nvmfcleanup
killprocess $nvmfpid
timing_exit fs_test
`
 steps to reproduce this issue:
 Run the updated test/nvmf/filesystem/filesystem.sh.

Result:
The script hung up or SPDK nvmf target crashed.

**Comment by dverkamp-intel** [comment]
The script above changes a couple of things compared to the current test:
- Malloc bdevs are changed to NVMe + lvol
- Nightly test builds a kernel source tree instead of touching a file

Does each of these changes have an associated Gerrit review?

Can you narrow down which part is causing the issue?

**Comment by wanqunintel** [comment]
@liang2yan Is there a patch review for it?

**Comment by liang2yan** [comment]
@wanqunintel , there is no patch to fix this issue. Only one patch is for the test scripts.
https://review.gerrithub.io/#/c/393299/

@dverkamp-intel ,please see the link(https://review.gerrithub.io/#/c/393299/
) for this script.  Actually the script didn't run into the building kernel source step. It failed at the mkpart step. I have used NVMe backend only. And the scripts passed. And I  used Malloc+LVOL backend and run the scripts again. The script also hung up.
I have added "-t all" in the nvmf target boot command line, some info will be printed:
<pre><code>
 15:42:42       # parted -s /dev/nvme0n1 mklabel msdos mkpart primary 0% 100%
rdma.c: 966:spdk_nvmf_rdma_request_process: *INFO*: Request 0x7f5748014800 entering state 1
rdma.c: 966:spdk_nvmf_rdma_request_process: *INFO*: Request 0x7f5748014800 entering state 5
request.c:  86:nvmf_trace_command: *INFO*: I/O cmd: opc 0x00 fuse 0 cid 54 nsid 1 cdw10 0x00000000
vbdev_lvol.c: 680:vbdev_lvol_submit_request: *INFO*: Vbdev request type 4 submitted
**vbdev_lvol.c: 700:vbdev_lvol_submit_request: *INFO*: lvol: unsupported I/O type 4**
</pre></code>
Then the script reported error:
Warning: Error fsyncing/closing /dev/nvme0n1: Input/output error

The script run into the trap handling function and hung up at "$rpc_py destroy_lvol_store -l lvs0" step.

**Comment by changpe1** [comment]
There are 2 steps to fix this issue, first, we need to set the NVMoF controller identity's volatile write cache bit based on the backend device can support FLUSH command or not, I have submitted a patch for this: https://review.gerrithub.io/#/c/402653/; second, there need to add FLUSH API support for lvol bdev layer, this should be another patch.

**Comment by changpe1** [comment]
Yes, I also reproduced the issue in my local test environment, the NVMe+LVOL backend failed with mkfs.xfs tests, ext4 and btrfs passed, I will continue debugging the issue.

**Comment by tomzawadzki** [comment]
@liang2yan Can you please see patch https://review.gerrithub.io/#/c/403306/ I believe this should resolve this issue.
